packages/grafana-data/src/types/datasource.ts

```diff
@@ -25,6 +25,15 @@ export type DataSourceQueryType<DSType> = DSType extends DataSourceApi<infer TQu
 // Utility type to extract the options type TOptions from a class extending DataSourceApi<TQuery, TOptions>
 export type DataSourceOptionsType<DSType> = DSType extends DataSourceApi<any, infer TOptions> ? TOptions : never;
 
+export type QueryRelatedDataProviders = {
+  [name in string]: QueryRelatedDataProvider<any>;
```

Comment:
is the key significant? I also assume it may be better than having a simple array but if we have it we should be clear what it should look like, like some id that will be used elsewhere or semi-user-readable etc.

Reply:
Totally could be an array and we could store data for each provider, but at some point we need some sort of identifier so a component can pick the correct data to render. We want it extendable but have to avoid clashes. We could also use an enum for - the components to render the data will be created internally. But if we allow creating plugins for Explore it'd need to be more dynamic.

Reply:
Question is what will be the identifier? Like for the user or the application. In the that if we had multiple usecases we will have to present a list of things user can run, either as a buttons or a dropdown or what ever ui with some user readable titles so user knows what it represents. At that point I assume the `getMetadata()` method would come into play as it could also give you some title and some more lengthy tooltip text or something like that. But at that point I still don't see where the name from this mapping will come into play. Or am I looking at this wrong?

Reply:
The only reason I added it was that once we have multiple providers, each component would need to read data from a specific provider. At the moment result of the provider is saved directly in ExploreItemState as `logsVolume` and this property is read from the state by LogsVolume component. With multiple providers, data will be aggregated in some sort of collection but then the rendering component would need to know which data pick from that collection and it could be done based on an identifier.

This means that the identifier should define the type of data that some Explore components can render. In this case type of the data would by "logs volume" and any data source that wants to use this feature needs to return provider of a same "type". 

So I was thinking we could add an enum of supported types.

Reply:
Ah ok interesting I think I get it now. One thing though is that as you said in other comment the provider can possibly represented multiple queries with different "types" right? In the sense that one provider could run 2 queries one for logs volume but other for some metrics for example. At that point this grouping key on provider breaks down a bit as the grouping should probably be done on the actual responses similarly to what we do with preferredVisualisationType right now which is property of the data frame.

Maybe that would be the way to go in the future. To have some more variable grouping key in the data frame which is not about the data type but semantic. Ie log volume is also metric but you want log volume in separate graph rather then have it together with some other metrics.

Reply:
Actually I didn't think about it this way. I thought about having multiple providers: each providing only one type of data, not single provider with multiple types of data if that makes sense.

By "multiple queries" I meant more something that is implementation details internal to the data provider. The result emitted by the provider is `TData` (e.g. `LogsVolume`) which in theory may not even contain data frames. For example for a given Prometheus query, the data provder could check if similar labels are in Loki data source and return information that logs are available for same PromQL query but in Loki.

By "multiple queries" I was thinking about what happens in logs volume provider - it executes multiple queries/requests (for each query row in Explore) and then merges results together into single DataFrame + extra information all wrapped in `LogsVolume` object. Of course, the data source could also, under the hood, execute only a single query (which theoretically could be possible by joining multiple loki queries together somehow), but the reducer/UI doesn't know that details. It just expects `LogsVolume` to be emitted.

Reply:
To maybe illustrate is a bit better I've changed it so the state now keeps collection of different data providers based on their type:

https://github.com/grafana/grafana/pull/39042/files#diff-18f2fb64eb8810155f6036e346139c7c224eb45ab9d436edf726bda7c3a58e29

Logs volume data is retrieved in https://github.com/grafana/grafana/pull/39042/files#diff-675f40aaa0e8180a3887e69c2969ba66e31f2558ed43436ae2f7efc0e580c9c5R274

```typescript
renderLogsVolume(width: number) {
    const { relatedData } = this.props;
    const logsVolume = relatedData[RelatedDataType.LogsVolume];

    return (
      <Collapse label="Logs volume" isOpen={true} loading={logsVolume?.isLoading}>
        {this.getLogsVolumeContent(width, logsVolume)}
      </Collapse>
    );
  }
```

Also here's an example of how it could be extended to show related metrics for given query: https://github.com/grafana/grafana/compare/ifrost/query-related-data-providers...ifrost/prototype/other-data-providers?expand=1 (also see the video in the PR description)